:has(:lang(~)) doesn't get invalidated

rniwa opened the pull request:
#### d66896dfcf4fa8f57a526003cfe8144e0360adda
<pre>
:has(:lang(~)) doesn&apos;t get invalidated
<a href="https://bugs.webkit.org/show_bug.cgi?id=243172">https://bugs.webkit.org/show_bug.cgi?id=243172</a>

Reviewed by Antti Koivisto.

Make container query work with :lang pseudo class.

This patch introduces ElementRareData::effectiveLang to store the language of each element,
and uses Style::PseudoClassChangeInvalidation to invalidate styles for :lang pseudo class.

* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-xhtml-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-xhtml.xhtml: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation.html: Added.

* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass): Use newly introduced effectiveLang.

* Source/WebCore/dom/Element.cpp:
(WebCore::effectiveLangFromAttribute):
(WebCore::Element::attributeChanged): Added the style invalidation logic for :lang.
(WebCore::Element::insertedIntoAncestor): Propagate the lang attribute from the parent.
(WebCore::Element::removedFromAncestor): Clear the effective lang when appropriate.
(WebCore::Element::effectiveLang const): Added.

* Source/WebCore/dom/Element.h:

* Source/WebCore/dom/ElementRareData.cpp:
(WebCore::SameSizeAsElementRareData):

* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::effectiveLang const): Added.
(WebCore::ElementRareData::setEffectiveLang): Added.
(WebCore::ElementRareData::useTypes const):

* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/style/PseudoClassChangeInvalidation.cpp:
(WebCore::Style::PseudoClassChangeInvalidation::computeInvalidation):
(WebCore::Style::PseudoClassChangeInvalidation::collectRuleSets):
* Source/WebCore/style/PseudoClassChangeInvalidation.h:
(WebCore::Style::PseudoClassChangeInvalidation::Value): Added.
(WebCore::Style::PseudoClassChangeInvalidation::PseudoClassChangeInvalidation): Added a new variant
which takes AnyValueTag as an argument.

* Tools/TestWebKitAPI/Tests/WebCore/DocumentOrder.cpp:
(TestWebKitAPI::createDocument): Initialize XMLNames via ProcessWarming.
</pre>

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3828,6 +3844,13 @@ AtomString Element::computeInheritedLanguage() const
     return document().contentLanguage();
 }
 
+AtomString Element::langForStyleResolution() const
```
I'd call this 'effectiveLang' or similar.

rniwa replied:
effectiveLangForStyleResolution? or just effectiveLang?

rniwa replied:
Done.

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3828,6 +3844,13 @@ AtomString Element::computeInheritedLanguage() const
+{
+    if (!hasRareData())
+        return nullAtom();
+    return elementRareData()->langForStyleResolution();
```
shouldn't this return document element lang if not explicitly set?

rniwa replied:
Fixed.

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3828,6 +3844,13 @@ AtomString Element::computeInheritedLanguage() const
+    return elementRareData()->langForStyleResolution();
+}
+
 Locale& Element::locale() const
 {
     return document().getCachedLocale(computeInheritedLanguage());
```
Can this use of computeInheritedLanguage() also be replaced with the new mechanism? Can all of them?

rniwa replied:
I'm gonna do that as a follow up.

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2485,6 +2516,20 @@ Node::InsertedIntoAncestorResult Element::insertedIntoAncestor(InsertionType ins
             CustomElementReactionQueue::enqueueConnectedCallbackIfNeeded(*this);
     }
 
+    [&]() {
```
why this?

rniwa replied:
I'm using it to have an early return.

rniwa committed:
:has(:lang(~)) doesn't get invalidated
https://bugs.webkit.org/show_bug.cgi?id=243172

Reviewed by Antti Koivisto.

Make container query work with :lang pseudo class.

This patch introduces ElementRareData::effectiveLang to store the language of each element,
and uses Style::PseudoClassChangeInvalidation to invalidate styles for :lang pseudo class.

* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-xhtml-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-xhtml.xhtml: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation.html: Added.

* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass): Use newly introduced effectiveLang.

* Source/WebCore/dom/Element.cpp:
(WebCore::effectiveLangFromAttribute):
(WebCore::Element::attributeChanged): Added the style invalidation logic for :lang.
(WebCore::Element::insertedIntoAncestor): Propagate the lang attribute from the parent.
(WebCore::Element::removedFromAncestor): Clear the effective lang when appropriate.
(WebCore::Element::effectiveLang const): Added.

* Source/WebCore/dom/Element.h:

* Source/WebCore/dom/ElementRareData.cpp:
(WebCore::SameSizeAsElementRareData):

* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::effectiveLang const): Added.
(WebCore::ElementRareData::setEffectiveLang): Added.
(WebCore::ElementRareData::useTypes const):

* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/style/PseudoClassChangeInvalidation.cpp:
(WebCore::Style::PseudoClassChangeInvalidation::computeInvalidation):
(WebCore::Style::PseudoClassChangeInvalidation::collectRuleSets):
* Source/WebCore/style/PseudoClassChangeInvalidation.h:
(WebCore::Style::PseudoClassChangeInvalidation::Value): Added.
(WebCore::Style::PseudoClassChangeInvalidation::PseudoClassChangeInvalidation): Added a new variant
which takes AnyValueTag as an argument.

Canonical link: https://commits.webkit.org/253530@main
